cpp/src/arrow/util/bit_util.h

```diff
@@ -827,6 +834,164 @@ Status InvertBitmap(MemoryPool* pool, const uint8_t* bitmap, int64_t offset,
 ARROW_EXPORT
 int64_t CountSetBits(const uint8_t* data, int64_t bit_offset, int64_t length);
 
+class ARROW_EXPORT Bitmap {
+ public:
+  Bitmap() = default;
+
+  Bitmap(std::shared_ptr<Buffer> buffer, int64_t offset, int64_t length)
+      : buffer_(std::move(buffer)), offset_(offset), length_(length) {}
+
+  Bitmap(const uint8_t* data, int64_t offset, int64_t length)
+      : buffer_(std::make_shared<Buffer>(data, BitUtil::BytesForBits(offset + length))),
+        offset_(offset),
+        length_(length) {}
+
+  Bitmap Slice(int64_t offset) const {
+    return Bitmap(buffer_, offset_ + offset, length_ - offset);
+  }
+
+  Bitmap Slice(int64_t offset, int64_t length) const {
+    return Bitmap(buffer_, offset_ + offset, length);
+  }
+
+  std::shared_ptr<BooleanArray> ToArray() const;
+
+  std::string ToString() const;
+
+  std::string Diff(const Bitmap& other) const;
+
+  bool GetBit(int64_t i) const { return BitUtil::GetBit(buffer_->data(), i + offset_); }
+
+  bool operator[](int64_t i) const { return GetBit(i); }
+
+  void SetBitTo(int64_t i, bool v) const {
+    BitUtil::SetBitTo(buffer_->mutable_data(), i + offset_, v);
+  }
+
+  /// \brief Visit bits of this bitmap either as bitset<N> or as array<Word, N>.
+  ///
+  /// If visiting by words, note that the every Word containing a bit
+  /// must be accessible (and initialized, or Valgrind will barf).
+  /// This is satisfied by most Buffers in Arrow space (which are allocated with
+  /// maximum alignment and padded to contain whole cache lines).
+  ///
+  /// TODO(bkietz) allow for early termination
+  template <size_t N, typename Visitor>
+  static void Visit(const Bitmap (&bitmaps)[N], Visitor&& visitor) {
+    VisitImpl(bitmaps, VisitedIs<internal::call_traits::argument_type<0, Visitor&&>>{},
+              std::forward<Visitor>(visitor));
+  }
+
+  const std::shared_ptr<Buffer>& buffer() const { return buffer_; }
+
+  /// offset of first bit relative to buffer().data()
+  int64_t offset() const { return offset_; }
+
+  /// number of bits in this Bitmap
+  int64_t length() const { return length_; }
+
+  /// string_view of all bytes which contain any bit in this Bitmap
+  util::basic_string_view<uint8_t> bytes() const {
+    auto byte_offset = offset_ / 8;
+    auto byte_count = BitUtil::CeilDiv(offset_ + length_, 8) - byte_offset;
+    return util::basic_string_view<uint8_t>(buffer_->data() + byte_offset, byte_count);
+  }
+
+  template <typename Word>
+  bool word_aligned() const {
+    auto words = this->words<Word>();
+    auto valid_addr = AsInt(buffer_->data());
+    auto words_addr = AsInt(words.data());
+    return words_addr >= valid_addr &&
+           words_addr + words.size() * sizeof(Word) <= valid_addr + buffer_->capacity();
+  }
+
+  /// string_view of all Words which contain any bit in this Bitmap
+  template <typename Word>
+  util::basic_string_view<Word> words() const {
```

Comment:
Could be helpful to give a schematic comment illustrating what's returned

Reply:
will do